### DISCUSSION TITLE
X86 Quantization on test_matmul.py?

### DISCUSSION THREAD

**Discussion by rtayl** [microsoft/triton-shared#168] [General]
Is anyone doing quantization on x86 for fp32 to int8, I've been looking at some examples and I'm trying to do this for the test_matmul.py in the triton_shared/python/examples dir but am not able to do it. Does triton even support int8 at all? For example, if I modify the MLIR file after --triton-to-linalg-experimental by changing f32 to i8 I run into translation errors when trying to lower to affine loops. If I change float32 to int8 at the source level for test_matmul.py I run into errors about int8 not being supported in triton_shared: RuntimeError: "normal_kernel_cpu" not implemented for 'Char'
. I have seen this: https://pytorch.org/blog/int8-quantization/ but could not get this to work for test_matmul.py.

**Comment by parsifal-47** [comment]
> Is anyone doing quantization on x86 for fp32 to int8, I've been looking at some examples and I'm trying to do this for the test_matmul.py in the triton_shared/python/examples dir but am not able to do it. Does triton even support int8 at all? For example, if I modify the MLIR file after --triton-to-linalg-experimental by changing f32 to i8 I run into translation errors when trying to lower to affine loops. If I change float32 to int8 at the source level for test_matmul.py I run into errors about int8 not being supported in triton_shared: RuntimeError: "normal_kernel_cpu" not implemented for 'Char' . I have seen this: https://pytorch.org/blog/int8-quantization/ but could not get this to work for test_matmul.py.

Hi, yes, Triton supports int8. Could you please share the error that you see? Maybe I can help fixing it.

**Reply by parsifal-47** [reply]
and if you can provide the exact test that you run it would be also helpful. Thank you!

**Reply by rtayl** [reply]
> and if you can provide the exact test that you run it would be also helpful. Thank you!

Thanks for replying. 
As I mentioned, the exact test can be found in triton_shared/python/examples/test_matmul.py. Also, as I mentioned above the exact error I see is: 
a = torch.randn((rows1, cols1), device=device, dtype=torch.int8)
E       RuntimeError: "normal_kernel_cpu" not implemented for 'Char'

test_matmul.py:155: RuntimeError
==================================== short test summary info ====================================
FAILED test_matmul.py::test_matmul - RuntimeError: "normal_kernel_cpu" not implemented for 'Char'
======================================= 1 failed in 2.04s =======================================

**Reply by parsifal-47** [reply]
It could be that this particular kernel needs modification, I see that the test for "dot" (aka matmul) contains "int8" clause here:
https://github.com/triton-lang/triton/blob/main/python/test/unit/language/test_core.py#L3088
Which means it should be supported by the language itself.
Let me check and get back to you!

**Reply by rtayl** [reply]
I tried grepping for "normal_kernel_cpu" but that returns nothing, so not sure what is reporting this error.

**Reply by parsifal-47** [reply]
I just looked, the error is generated by pytorch, not the triton itself:
```
>       a = torch.randn((rows1, cols1), device=device, dtype=torch.int8)
E       RuntimeError: "normal_kernel_cpu" not implemented for 'Char'
```
and if I look at how such tensors are getting generated in triton tests, they do numpy and convert to torch.
However, when I fix that, I see it crashing in runtime. I can take a look at why "dot" with int8 is crashing and get back to you.

**Reply by rtayl** [reply]
> It could be that this particular kernel needs modification, I see that the test for "dot" (aka matmul) contains "int8" clause here: https://github.com/triton-lang/triton/blob/main/python/test/unit/language/test_core.py#L3088 Which means it should be supported by the language itself. Let me check and get back to you!

By the way, if it wasn't explicit above, I merely did a replace of "float32" with "int8" in that test_matmul.py example

**Reply by parsifal-47** [reply]
> By the way, if it wasn't explicit above, I merely did a replace of "float32" with "int8" in that test_matmul.py example

yes, that's exactly what I did in my test, it hit the error with `torch.rand` first and then hit something internal once I fixed the `rand`.

**Reply by rtayl** [reply]
> > By the way, if it wasn't explicit above, I merely did a replace of "float32" with "int8" in that test_matmul.py example
> 
> yes, that's exactly what I did in my test, it hit the error with `torch.rand` first and then hit something internal once I fixed the `rand`.

Can you tell me how you fixed this in torch.rand?

**Reply by parsifal-47** [reply]
sure:
```
    a = torch.randint(low=-128, high=128, size=(rows1, cols1), device='cpu', dtype=torch.int8)
    b = torch.randint(low=-128, high=128, size=(rows2, cols2), device='cpu', dtype=torch.int8)
```

**Reply by rtayl** [reply]
> sure:
> 
> ```
>     a = torch.randint(low=-128, high=128, size=(rows1, cols1), device='cpu', dtype=torch.int8)
>     b = torch.randint(low=-128, high=128, size=(rows2, cols2), device='cpu', dtype=torch.int8)
> ```

Shouldn't this be -128, 127?

**Reply by rtayl** [reply]
> > By the way, if it wasn't explicit above, I merely did a replace of "float32" with "int8" in that test_matmul.py example
> 
> yes, that's exactly what I did in my test, it hit the error with `torch.rand` first and then hit something internal once I fixed the `rand`.

Is this the internal error you are getting?:

AssertionError('Loop-carried variable accumulator has initial type <[16, 16], int8> but is re-assigned to <[16, 16], int32> in loop! Please make sure that the type stays consistent.')

It looks like tl.dot is perhaps converting the int8 to an int32 and returning that type back to accumulator in the loop. Is that what you've seen?

EDIT; Tried to add an ".to(int8)" to tl.dot call but that got errors. This seemed to get further still error so not sure this is the right thing here.

**Reply by parsifal-47** [reply]
> > sure:
> > ```
> >     a = torch.randint(low=-128, high=128, size=(rows1, cols1), device='cpu', dtype=torch.int8)
> >     b = torch.randint(low=-128, high=128, size=(rows2, cols2), device='cpu', dtype=torch.int8)
> > ```
> 
> Shouldn't this be -128, 127?

similarly to range() in python, the upper bound is exclusive

**Reply by parsifal-47** [reply]
> Is this the internal error you are getting?:
> 
> AssertionError('Loop-carried variable accumulator has initial type <[16, 16], int8> but is re-assigned to <[16, 16], int32> in loop! Please make sure that the type stays consistent.')
> 
> It looks like tl.dot is perhaps converting the int8 to an int32 and returning that type back to accumulator in the loop. Is that what you've seen?
> 
> EDIT; Tried to add an ".to(int8)" to tl.dot call but that got errors. This seemed to get further still error so not sure this is the right thing here.

yes, I saw this error and I tried some simple workarounds like changing acc to int32, but it looks like something wrong is happening internally, need to look at internals

**Reply by parsifal-47** [reply]
I have found that tt.dot wasn't implemented properly, here is the fix:
https://github.com/microsoft/triton-shared/pull/170

**Reply by rtayl** [reply]
> I have found that tt.dot wasn't implemented properly, here is the fix: #170

This doesn't seem to have fixed the issue, I am still getting the same error.

**Reply by parsifal-47** [reply]
> a = torch.randint(low=-128, high=128, size=(rows1, cols1), device='cpu', dtype=torch.int8)
>     b = torch.randint(low=-128, high=128, size=(rows2, cols2), device='cpu', dtype=torch.int8)

sorry, I forgot to follow up, the kernel now works for me, below are the changes, if these changes do not work, try to purge triton cache with "rm -rf ~/.triton/cache/*"

```
-    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.float32)
+    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=tl.int32)

-    c = accumulator.to(tl.float32)
+    c = accumulator.to(tl.int32)

-    a = torch.randn((rows1, cols1), device=device, dtype=torch.float32)
-    b = torch.randn((rows2, cols2), device=device, dtype=torch.float32)

+    a = torch.randint(low=-128, high=128, size=(rows1, cols1), device='cpu', dtype=torch.int8)
+    b = torch.randint(low=-128, high=128, size=(rows2, cols2), device='cpu', dtype=torch.int8)
```

please let me know if it works for you, thank you!
